File: justice-dashboard/scripts/update-pdf-links.js
#!/usr/bin/env node

/**
 * PDF Link Updater - JavaScript Version
 * Updates PDF links and bookmarks for Justice Dashboard
 *
 * Usage:
 *   node scripts/update-pdf-links.js <input.pdf> [output.pdf]
 *   npm run update-pdf-links <input.pdf> [output.pdf]
 */

const fs = require('fs');
const path = require('path');
const { spawn } = require('child_process');

// Configuration
const DEFAULT_OUTPUT = 'MCL, Federal Law- Misconduct Analysis (2).pdf';
const COMMON_PDF_LOCATIONS = ['server/uploads', 'uploads', '.'];

// Colors for console output
const colors = {
  reset: '\x1b[0m',
  red: '\x1b[31m',
  green: '\x1b[32m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  cyan: '\x1b[36m',
  gray: '\x1b[90m',
};

function log(message, color = 'reset') {
  console.log(`${colors[color]}${message}${colors.reset}`);
}

function findPDFFiles() {
  const foundPdfs = [];

  for (const location of COMMON_PDF_LOCATIONS) {
    if (fs.existsSync(location)) {
      try {
        const files = fs.readdirSync(location);
        const pdfFiles = files.filter(file =>
          file.toLowerCase().endsWith('.pdf')
        );

        for (const file of pdfFiles) {
          const fullPath = path.join(location, file);
          const stats = fs.statSync(fullPath);
          foundPdfs.push({
            path: fullPath,
            name: file,
            size: Math.round(stats.size / 1024), // Size in KB
          });
        }
      } catch (error) {
        // Ignore permission errors
      }
    }
  }

  return foundPdfs;
}

function displayPDFFiles(pdfs) {
  if (pdfs.length === 0) {
    log('\n[ERROR] No PDF files found in common locations:', 'red');
    log('Expected locations:', 'yellow');
    log('   • server/uploads/', 'gray');
    log('   • uploads/', 'gray');
    log('   • current directory (.)', 'gray');
    log('\nTroubleshooting steps:', 'yellow');
    log('1. Check if PDFs exist in any of the above folders', 'gray');
    log('2. Ensure you have read permissions for these locations', 'gray');
    log('3. Try providing the full absolute path to your PDF file:', 'gray');
    log('   Example: C:\\Users\\ssped\\Documents\\input.pdf', 'gray');
    log('\nUsage:', 'yellow');
    log('  node scripts/update-pdf-links.js <input.pdf> [output.pdf]', 'gray');
    return;
  }

  log(`\n[FOUND] ${pdfs.length} PDF file(s):`, 'green');
  for (let i = 0; i < Math.min(pdfs.length, 10); i++) {
    const pdf = pdfs[i];
    log(`   ${i + 1}. ${pdf.path} (${pdf.size}KB)`, 'cyan');
  }

  if (pdfs.length > 10) {
    log(`   ... and ${pdfs.length - 10} more files`, 'gray');
  }

  log(
    '\nTIP: Copy and paste one of the paths above, or provide a custom path.',
    'yellow'
  );
}

function validateInputFile(inputPath) {
  if (!inputPath) {
    log('[ERROR] Input file path is required.', 'red');
    return false;
  }

  if (!fs.existsSync(inputPath)) {
    log(`[ERROR] Input file not found: ${inputPath}`, 'red');
    log('Please check the file path and try again.', 'gray');
    return false;
  }

  if (!inputPath.toLowerCase().endsWith('.pdf')) {
    log('[WARNING] Input file does not have a .pdf extension.', 'yellow');
  }

  return true;
}

function runPythonScript(inputPath, outputPath) {
  return new Promise((resolve, reject) => {
    const pythonScript = path.join(__dirname, '..', 'update_pdf_links.py');

    // Check if Python script exists
    if (!fs.existsSync(pythonScript)) {
      log(`[ERROR] Python script not found: ${pythonScript}`, 'red');
      log(
        'Please ensure update_pdf_links.py exists in the project root.',
        'gray'
      );
      reject(new Error('Python script not found'));
      return;
    }

    log('\nProcessing PDF...', 'green');
    log(`Input:  ${inputPath}`, 'gray');
    log(`Output: ${outputPath}`, 'gray');
    log('');

    const python = spawn('python', [pythonScript, inputPath, outputPath], {
      stdio: 'inherit',
    });

    python.on('close', code => {
      if (code === 0) {
        log('\n[SUCCESS] PDF processing completed successfully!', 'green');
        log(`Output saved as: ${outputPath}`, 'cyan');

        // Show file info if it exists
        if (fs.existsSync(outputPath)) {
          const stats = fs.statSync(outputPath);
          const fileSizeKB = Math.round(stats.size / 1024);
          log(`File size: ${fileSizeKB} KB`, 'gray');
        }
        resolve();
      } else {
        log('\n[ERROR] PDF processing failed!', 'red');
        log('Check that:', 'yellow');
        log('   • Input file path is correct', 'gray');
        log('   • Python and PyPDF2 are installed', 'gray');
        log('   • You have write permissions for the output location', 'gray');
        reject(new Error(`Python script exited with code ${code}`));
      }
    });

    python.on('error', error => {
      log('\n[ERROR] Failed to start Python process:', 'red');
      log(error.message, 'gray');
      log('\nEnsure Python is installed and available in your PATH.', 'yellow');
      reject(error);
    });
  });
}

async function main() {
  log('PDF Link Updater - JavaScript Version', 'cyan');
  log('====================================', 'cyan');

  const args = process.argv.slice(2);
  let inputPath = args[0];
  let outputPath = args[1] || DEFAULT_OUTPUT;

  // If no input provided, show available PDFs and exit
  if (!inputPath) {
    log('\nScanning for PDF files in common locations...', 'yellow');
    const pdfs = findPDFFiles();
    displayPDFFiles(pdfs);

    log('\nUsage:', 'yellow');
    log('  node scripts/update-pdf-links.js <input.pdf> [output.pdf]', 'gray');
    log('  npm run update-pdf-links <input.pdf> [output.pdf]', 'gray');
    log('\nExample:', 'yellow');
    log(
      '  node scripts/update-pdf-links.js "server/uploads/document.pdf"',
      'gray'
    );
    log(
      '  npm run update-pdf-links "C:/Documents/legal.pdf" "Updated_Legal.pdf"',
      'gray'
    );

    process.exit(1);
  }

  // Validate input file
  if (!validateInputFile(inputPath)) {
    process.exit(1);
  }

  // Show current configuration
  log('\nConfiguration:', 'blue');
  log(`  Input:  ${inputPath}`, 'gray');
  log(`  Output: ${outputPath}`, 'gray');

  try {
    await runPythonScript(inputPath, outputPath);
    log('\n✓ PDF link update process completed!', 'green');
  } catch (error) {
    log('\n✗ PDF link update process failed!', 'red');
    process.exit(1);
  }
}

// Handle unhandled promise rejections
process.on('unhandledRejection', error => {
  log('\n[ERROR] Unhandled promise rejection:', 'red');
  log(error.message, 'gray');
  process.exit(1);
});

// Handle SIGINT (Ctrl+C)
process.on('SIGINT', () => {
  log('\n\nOperation cancelled by user.', 'yellow');
  process.exit(0);
});

// Run the main function
if (require.main === module) {
  main().catch(error => {
    log('\n[ERROR] Script execution failed:', 'red');
    log(error.message, 'gray');
    process.exit(1);
  });
}

module.exports = {
  findPDFFiles,
  validateInputFile,
  runPythonScript,
  DEFAULT_OUTPUT,
  COMMON_PDF_LOCATIONS,
};

File: justice-dashboard/update_pdf_links.py
#!/usr/bin/env python3
"""
PDF Link Updater and Bookmark Creator
Applies hyperlink corrections and adds bookmarks to PDF files

Requirements:
    pip install PyPDF2

Usage:
    python update_pdf_links.py input.pdf [output.pdf]
"""

import sys
import os
from pathlib import Path

# Safe import with proper error handling
try:
    import PyPDF2
except ImportError:
    print("\n❌ ERROR: PyPDF2 is not installed.")
    print("\nTo install the required dependency, run:")
    print("    pip install PyPDF2")
    print("\nOr if using conda:")
    print("    conda install -c conda-forge pypdf2")
    print("\nThen try running this script again.")
    sys.exit(1)

def update_pdf_links(input_file, output_file=None):
    """
    Updates PDF links and adds bookmarks
    
    Args:
        input_file (str): Path to input PDF
        output_file (str): Path to output PDF (optional)
    """
    
    if not os.path.exists(input_file):
        print(f"\nError: Input file '{input_file}' not found.")
        print("\nTroubleshooting steps:")
        print("1. Check if the file exists in these locations:")
        print("   • server/uploads/")
        print("   • uploads/")
        print("   • current directory")
        print("2. Verify file permissions")
        print("3. Try using the full absolute path")
        print("\nExample: python update_pdf_links.py C:\\Users\\ssped\\Documents\\input.pdf")
        return False
    
    if output_file is None:
        base_name = Path(input_file).stem
        output_file = f"{base_name}_updated.pdf"
    
    try:
        print(f"Processing PDF: {input_file}")
        
        # Read the input PDF
        with open(input_file, 'rb') as file:
            reader = PyPDF2.PdfReader(file)
            writer = PyPDF2.PdfWriter()
            
            print(f"Found {len(reader.pages)} pages")
            
            # Copy pages and update annotations
            for page_num, page in enumerate(reader.pages):
                print(f"Processing page {page_num + 1}...")
                
                # Check for annotations and update links
                if '/Annots' in page:
                    annotations = page['/Annots']
                    for annot_ref in annotations:
                        try:
                            annot = annot_ref.get_object()
                            if annot and '/A' in annot:
                                action = annot['/A']
                                if '/URI' in action:
                                    old_uri = str(action['/URI'])
                                    # Update specific patterns
                                    if 'old-link.pdf' in old_uri:
                                        action[PyPDF2.generic.NameObject('/URI')] = PyPDF2.generic.TextStringObject(output_file)
                                        print(f"Updated link: {old_uri} -> {output_file}")
                        except Exception as e:
                            print(f"Warning: Could not process annotation on page {page_num + 1}: {e}")
                
                writer.add_page(page)
            
            # Add bookmarks for legal document structure
            print("Adding bookmarks...")
            
            # Calculate approximate bookmark positions based on page count
            total_pages = len(reader.pages)
            
            if total_pages >= 3:
                bookmark_positions = {
                    "1. Executive Summary": 0,
                    "2. Police Misconduct Analysis": max(1, total_pages // 4),
                    "3. CPA Misconduct Documentation": max(2, total_pages // 2),
                    "4. CPS Misconduct Evidence": max(3, (3 * total_pages) // 4),
                    "5. Legal Conclusions": max(4, total_pages - 2)
                }
                
                for title, page_num in bookmark_positions.items():
                    if page_num < total_pages:
                        writer.add_outline_item(title, page_num)
                        print(f"Added bookmark: {title} (page {page_num + 1})")
            
            # Write the updated PDF
            with open(output_file, 'wb') as output:
                writer.write(output)
        
        print(f"Successfully updated PDF saved as: {output_file}")
        return True
        
    except Exception as e:
        print(f"Error processing PDF: {e}")
        return False

def main():
    """Main function to handle command line arguments"""
    
    if len(sys.argv) < 2:
        print("Usage: python update_pdf_links.py <input_file> [output_file]")
        print("Example: python update_pdf_links.py old.pdf 'MCL, Federal Law- Misconduct Analysis (2).pdf'")
        return
    
    input_file = sys.argv[1]
    output_file = sys.argv[2] if len(sys.argv) > 2 else None
    
    # Default output name if not specified
    if output_file is None:
        output_file = "MCL, Federal Law- Misconduct Analysis (2).pdf"
    
    success = update_pdf_links(input_file, output_file)
    
    if success:
        print("\nPDF processing completed successfully!")
        print(f"Input: {input_file}")
        print(f"Output: {output_file}")
    else:
        print("PDF processing failed.")
        sys.exit(1)

if __name__ == "__main__":
    main()
